File: app/src/main/java/com/robog/persistentlog/CrashHandler.java
package com.robog.persistentlog;

import com.robog.loglib.SLog;

/**
 * Created by yuxingdong on 2018/6/25.
 */
public class CrashHandler implements Thread.UncaughtExceptionHandler {

    private static final String TAG = "CrashHandler";
    private static Thread.UncaughtExceptionHandler sDefaultCrashHandler;
    private static CrashHandler sCrashHandler;

    private CrashHandler() {
        sDefaultCrashHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(this);
    }

    public static CrashHandler init() {
        if (sCrashHandler == null) {
            sCrashHandler = new CrashHandler();
        }
        return sCrashHandler;
    }

    @Override
    public void uncaughtException(Thread thread, Throwable t) {

        SLog.crash(TAG, t);

        if (sDefaultCrashHandler != null) {
            sDefaultCrashHandler.uncaughtException(thread, t);
        }
    }
}

File: sample/src/main/java/com/robog/persistentlog/CrashHandler.kt
package com.robog.persistentlog

import com.robog.loglib.SLog

/**
 * Created by yuxingdong on 2018/6/29.
 */
class CrashHandler private constructor() : Thread.UncaughtExceptionHandler {

    private val defaultCrashHandler: Thread.UncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()

    companion object {
        private const val TAG = "CrashHandler"

        private var instance: CrashHandler? = null

        fun init() = instance ?: CrashHandler().also { instance = it }
    }

    override fun uncaughtException(t: Thread?, e: Throwable?) {
        SLog.crash(TAG, e)
        defaultCrashHandler.uncaughtException(t, e)
    }
}